api/joinserver.proto

```diff
@@ -172,9 +137,9 @@ service JsEndDeviceRegistry {
     };
   };
 
-  // Provision returns end devices that are provisioned using the given vendor-specific data.
-  // The devices are not set in the registry.
-  rpc Provision(ProvisionEndDevicesRequest) returns (stream EndDevice) {
+  // This method is deprecated; use EndDeviceTemplateConverter service instead.
+  // TODO: Remove (https://github.com/TheThingsNetwork/lorawan-stack/issues/999)
+  rpc Provision(EndDeviceIdentifiers) returns (stream EndDevice) {
```

Comment:
Changing the signature still breaks the API guarantee, right ?

Reply:
As long as the proto is compatible, i.e. the field tags are of the same types, we're good.

Both `ProvisionEndDevicesRequest` and `EndDeviceIdentifiers` contains

```proto
ApplicationIdentifiers application_ids = 1
```

We need this one for the gRPC gateway, otherwise `google.protobuf.Empty` is cleanest.

I know, this is a hack and we'll remove it in V5

Reply:
Yeah I made a mistake here; the field tags of `ProvisionEndDevicesRequest` and `EndDeviceIdentifiers` are not all compatible. Will revert partly.

Reply:
https://github.com/TheThingsNetwork/lorawan-stack/pull/1016/commits/0accda90428bdefd9c456f637a327fb124512e87